File: BOJ/1874.cpp
#include <stdio.h>
#include <algorithm>
#include <stack>
#include <string.h>
using namespace std;
int n, p;
int arr[100001];
char c[250001];
stack <int> st;
int k = 1;
int main() {
	scanf("%d", &n);
	for (int i = 1; i <= n; i++) { scanf("%d", &arr[i]); }
	st.push(0);
	int Max = n;
	for (int i = 1; i <= n; i++) {
		int num = arr[i];
		while (num != st.top() && k <= Max) {
			st.push(k);
			c[p++] = '+';
			k++;
		}
		while (num == st.top()) {
			st.pop();
			c[p++] = '-';
			//if (st.empty()) break;
		}
	}
	if (st.top() == 0) {
		for (int i = 0; i < p; i++) {
			printf("%c\n", c[i]);
		}
	}
	else printf("NO\n");
	return 0;
}

File: BOJ(Java 공부용)/1874.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.ArrayList;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int[] gox = {0, 1, -1, 0, 1, 1, -1, -1};
    static int[] goy = {1, 0, 0, -1, 1, -1, 1, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = Integer.parseInt(br.readLine());
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        stack.add(0);
        int max = n, k = 1;
        ArrayList<Character> v = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            int num = arr[i];
            while (num != stack.peekLast() && k <= max) {
                stack.add(k++);
                v.add('+');
            }
            while (num == stack.peekLast()) {
                stack.pollLast();
                v.add('-');
            }
        }
        if (stack.peekLast() == 0) {
            for (int i = 0; i < v.size(); i++) {
                bw.write(v.get(i) + "\n");
            }
            bw.flush();
            bw.close();
        } else System.out.println("NO");
    }
}